### ISSUE TITLE
GUI hovers obscure touches

### ISSUE THREAD

**Issue by singalen** [singalen/wesnoth#2]
Instead of reacting to touches as clicks, GUI most times shows hover hints.
Sliders sometimes stop working 8(

**Comment by singalen** [comment]
Got the zucker! 
* `distributor.hpp`, see `mouse_button_left`. Create the same for touches.
* in `handler.cpp`, uncomment `if (event.motion.which != SDL_TOUCH_MOUSEID)` to stop handling touch events as mouse (will it break something else?)

**Comment by singalen** [comment]
[00:01] sinda: Another question: can anyone tell what function in listbox is responsible for list items recoloring when hovered?
I cannot find it after who knows which attempt 8-\
[00:03] Vultraz: I know, but not sure how best to explain it..
[00:04] sinda: Just point me to some lines in the code, I'll figure it out.
[00:04] Vultraz: https://github.com/wesnoth/wesnoth/blob/master/src/gui/widgets/toggle_panel.cpp#L207-L214
[00:05] Vultraz: that's bound to a MOUSE_ENTER event
[00:05] Vultraz: it sets the FOCUSED state
[00:05] Vultraz: which corresponds to a state defined in the toggle_panel WML
[00:06] Vultraz: that state definition defines the coloring

**Comment by singalen** [comment]
01:17] sinda: A question to someone with understanding of event distribution and tooltips.
Is it possible to close a tooltip on sdl_mouse_down - I do it in mouse_button<T>::signal_handler_sdl_button_down() with a mouse_leave() call - and continue handling the event further?
I'm afraid that the tooltip, being a modeless_dialog, is not closed yet, because NOTIFY_REMOVE_TOOLTIP is not handled yet - and thus swallows the sdl_mouse_down event.

It was mitigated by mouse_move events - we had the time to close a tooltip. But touch screen doesn't receive mouse_move when finger is up.
So far, I can think only about SDL_PushEvent(), or completely disabling the tooltips (it helps).
[01:18] sinda: I'm not completely sure in this analysis.
[01:21] jyrkive: Yes, it's possible. If you leave the handled parameter to false, you don't stop event propagation.
[01:26] sinda: Thanks. I'm not sure if it's true for sdl_event_handler::mouse() that is located down the stack, but I'll see if there is a place where handled helps.

02:02] sinda: @jyrkive I'm afraid in the described situation with tooltips there exists no place in the code where just 
It's about multiple windows interaction and events::pump().
Anyway, I'll just remove tooltips for touch screens, and that will do it for now.
[02:04] sinda: May I rant a little.
I felt there's something fishy with the event distribution design where each new window is a recursive show() call that calls events::pump() on its own, but couldn't immediately explain why, except for "nobody does that" and "too much overhead".
Now I finally found what was wrong with tooltips, lost mouse events and these recursive windows.
Tooltip is a dialog, and when it receives a mouse click and closes, the event is still considered handled by its dispatcher, because it's in the call stack's state, and nothing is going to change it untill the call stack unwinds.
It is mitigated by mouse_move events on PC - we have the time to close a tooltip. But touch screen doesn't receive mouse_move when finger is up.
Well, anyway.

**Comment by singalen** [comment]
tl;dr: Just remove tooltips on iOS. Touchscreen laptops will wait.
